proposal/WarClans612/proposal.rst

```diff
@@ -0,0 +1,154 @@
+in memory. For small enough graph, it is usually stored as flattened grid which is efficient in its
+I/O usage with sacrifice of storage size.
+
+Manipulating graph usually means dealing with matrix computation which is quite expensive to do.
+Therefore, there is a demand to improve efficiency in dealing with this kind of graph data. Many real
+world problems do not consists of dense matrix due to its nature.
+
+Example for this usage is for linguistics research. Since natural language often lends itself well
+of discrete structures. We can decompose semantics into tree-based structures or hierarchical graph.
+Another way to represents syntax for natural language can be modeled with directed acyclic graphs. The
+advantage of using graph is added by the fact that visualization if the tasks could help better understanding
+for the topic. Currently, mathematics for natural language related tasks has been used especially in
+artificial intelligence research, such as WordNet, VerbNet, TextGraphs, and etc.
+
+Math Behind the Problem
+-----------------
+
+Matrix calculation is the core math problem for graph manipulation. In this library, we would like
+extend support for matrix calculation especially for sparse matrix.
+
+There is 2 main methods to store sparse matrix, which is CSC (Compressed Sparse Column) and CSR 
+(Compressed Sparse Row). These 2 methods are quite similar except that values in CSC are read first by
+column, a row index is stored for each value, and column pointers are stored, while values in CSR are
+read first by row, a column index is stored for each value, and row pointers are stored. 
+
+Numerical Method to be Applied
+-----------------
+
+The main idea is to store sparse matrix using 2 supported method (CSC and CSR), and providing
+library that is able to do calculation with those 2 storage method. The graph class will then
+inherit from the matrix class to be used as further extension. However, in this project, we will
+focus more of feature implementation for the matrix rather than the graph.
+
+Perspective users
+=================
+
+- Research focus operation related to sparse matrix
+- Data analytic team that operates with huge nodes in graph
+- Company with huge amount of graph relation data collection that would like to decrease its storage usage
+
+System architecture
+===================
+
+Input:
+
+- Graph can be initialized as an empty graph
+- Provided method to add nodes into the graph
+- Provided method to add edges into the graph
+
+Output:
+
+- Provided method to store the matrix into a text file
+- Provided method to manipulate the matrix
+
+Modularization
+-----------------
+
+- SparseMatrix Class
+    - SparseMatrix Manipulation
+    - Storage
+    - Computation
+- Graph Class (inherit from Matrix Class):
+    - Graph Manipulation
+    - Storage (similar to matrix, graph can be represented as matrix)
+- Testing Module
+- Config Module
```

Comment:
Perhaps you consider a global configuration object.  It may work, but will impose limitation.  I think it's OK to take the global approach for a term project, but we can discuss more.

Reply:
I am thinking of letting the global configuration object as the default parameter for the object. In my experience, it is sometimes hard to find the value of variable after writing long code. Or do you have better solution for this parts? Thanks.

Reply:
ATM I think it's the way to go.  It shouldn't show milage in the project, but if you want to extend the project for a large scope, it may need improvements.